src/server/img_storage.rs

```diff
@@ -0,0 +1,230 @@
+use actix_http::http::HeaderValue;
+use actix_web::http::uri;
+use cloud_storage::{
+    bucket::{Binding, IamPolicy, IamRole, StandardIamRole},
+    Bucket, Object,
+};
+use serde::Deserialize;
+
+use crate::error::{HandlerErrorKind, HandlerResult};
+use crate::settings::Settings;
+
+#[derive(Clone, Debug, Deserialize)]
+pub struct StorageSettings {
+    project_name: String,
+    bucket_name: String,
+    endpoint: String,
+}
+
+impl Default for StorageSettings {
+    fn default() -> Self {
+        Self {
+            project_name: "secondary-project".to_owned(),
+            bucket_name: "moz-contile-test-jr".to_owned(),
+            endpoint: "https://storage.googleapis.com".to_owned(),
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct StoreImage {
+    // bucket isn't really needed here, since `Object` stores and manages itself,
+    // but it may prove useful in future contexts.
+    //
+    // bucket: Option<cloud_storage::Bucket>,
+    settings: StorageSettings,
+}
+
+#[derive(Debug)]
+pub struct StoreResult {
+    pub url: uri::Uri,
+    pub hash: String,
+    pub object: Object,
+}
+
+// TODO: Reduce all the `Internal` errors to more specific storage based ones
+
+impl StoreImage {
+    pub async fn create(settings: &Settings) -> HandlerResult<Self> {
+        let sset = settings.storage.clone();
+        // TOOD: Validate bucket name?
+        // https://cloud.google.com/storage/docs/naming-buckets
+        dbg!("Try creating bucket...");
+        let bucket = match Bucket::create(&cloud_storage::NewBucket {
+            name: sset.bucket_name.clone(),
+            ..Default::default()
+        })
+        .await
+        {
+            Ok(v) => v,
+            Err(cloud_storage::Error::Google(ger)) => {
+                if ger.errors_has_reason(&cloud_storage::Reason::Conflict) {
+                    dbg!("Already exists", &sset.bucket_name);
+                    // try fetching the existing bucket.
+                    match Bucket::read(&sset.bucket_name).await {
+                        Ok(_v) => {
+                            return Ok(Self {
+                                // bucket: Some(v),
+                                settings: sset,
+                            });
+                        }
+                        Err(e) => {
+                            return Err(HandlerErrorKind::Internal(format!(
+                                "Could not read bucket {:?}",
+                                e
+                            ))
+                            .into())
+                        }
+                    }
+                } else {
+                    return Err(HandlerErrorKind::Internal(format!(
```

Comment:
nit: Switch from `HandlerErrorKind::Internal` to `HandlerError::internal` to save a lil boilerplate (all the `into()`'s can go away)